Enhance Trim Deploy Documentation (#20135)
dotnet/docs: docs/core/deploying/trim-self-contained.md

---
# Trim self-contained deployments and executables

> The globally installed runtime and framework libraries has been the most successful runtime deployment model since the inception of .NET. In this scenario, the application developer will only bundle the application and third-party assemblies with the expectation that the .NET runtime and framework will be available in the client machine. This deployment model will continue to be the dominant one in .NET Core as well but there will be other application deployment models targeted at app developers where the above model is not the best one. One such one is targeted for some application developers that want to publish a [self-contained application](runtime-patch-selection.md) where the .NET Core runtime and framework are bundled together with the application.

Reviewer: Rephrasing and fixing grammar

Linking to base deployment overview since it hasn't been updated from runtime-dependent to framework-dependent yet

```suggestion
The [framework-dependent deployment model](index.md) has been the most successful deployment model since the inception of .NET. In this scenario, the application developer bundles only the application and third-party assemblies with the expectation that the .NET runtime and framework libraries will be available in the client machine. This deployment model continues to be the dominant one in .NET Core as well but there are some scenarios where the framework-dependent model is not optimal. The alternative is to publish a [self-contained application](index.md), where the .NET Core runtime and framework are bundled together with the application and third-party assemblies.
```

Reviewer 2: https://github.com/dotnet/docs/pull/20165

> Trim self-contained deployment model is a specialized version of self-contained deployment model and is optimized for size that is a critical requirement for some client-side scenarios like Blazor applications. Depending on the complexity of the application, only a subset of the framework assemblies are required to run the application. These unused parts of the library are unnecessary and can be trimmed from the packaged application. However, there is a risk that the build time analysis of the application can cause failures at runtime, due to not being able to reliably analyze various problematic code patterns (largely centered on reflection use). So this deployment model will be offered as a preview feature where the build time analysis engine will provide warnings to the developer of code patterns that are problematic with the expectation that these code patterns will be fixed. Where possible, it is encouraged to move any runtime reflection dependencies in your application to build time by using code that meets the same requirements.

Reviewer: Rephrasing, review for accuracy. It isn't clear, however, what "the same requirements" are that are mentioned at the end.

```suggestion
The trim-self-contained deployment model is a specialized version of the self-contained deployment model that is optimized to reduce deployment size. Minimizing deployment size is a critical requirement for some client-side scenarios like Blazor applications. Depending on the complexity of the application, only a subset of the framework assemblies are required to run the application. These unused parts of the library are unnecessary and can be trimmed from the packaged application. However, there is a risk that the build time analysis of the application can cause failures at runtime, due to not being able to reliably analyze various problematic code patterns (largely centered on reflection use). Because reliability can't be guaranteed, this deployment model is offered as a preview feature. The build time analysis engine provides warnings to the developer of code patterns that are problematic, with the expectation that these code patterns will be fixed. Where possible, we recommend that you move any runtime reflection dependencies in your application to build time by using code that meets the same requirements.
```

> The trim mode for the applications can be configured via the TrimMode and will default (`copyused`) to bundle assemblies that are used in the application. Blazor applications will use a more aggressive mode (`link`) that will trim unused code within assemblies. Trim analysis warnings give information on code patterns where a full dependency analysis was not possible. These warnings are suppressed by default and can be turned on by setting the flag, `SuppressTrimAnalysisWarnings`, to false. More information on the trim options available can be found at the [ILLinker page](https://github.com/mono/linker/blob/master/docs/illink-options.md).

Reviewer: minor rephrasing

```suggestion
The trim mode for applications can be configured via the `TrimMode`. The default (`copyused`) is to bundle assemblies that are used in the application. Blazor applications will use a more aggressive mode (`link`) that will trim unused code within assemblies. Trim analysis warnings give information on code patterns where a full dependency analysis was not possible. These warnings are suppressed by default and can be turned on by setting the flag `SuppressTrimAnalysisWarnings` to false. More information on the trim options available can be found at the [ILLinker page](https://github.com/mono/linker/blob/master/docs/illink-options.md).
```

> The trim mode for the applications can be configured via the TrimMode and will default (`copyused`) to bundle assemblies that are used in the application. Blazor applications will use a more aggressive mode (`link`) that will trim unused code within assemblies. Trim analysis warnings give information on code patterns where a full dependency analysis was not possible. These warnings are suppressed by default and can be turned on by setting the flag, `SuppressTrimAnalysisWarnings`, to false. More information on the trim options available can be found at the [ILLinker page](https://github.com/mono/linker/blob/master/docs/illink-options.md).

Reviewer 3: ```suggestion
The trim mode for the applications can be configured via the TrimMode and will default (`copyused`) to bundle assemblies that are used in the application. Blazor WebAssembly applications will use a more aggressive mode (`link`) that will trim unused code within assemblies. Trim analysis warnings give information on code patterns where a full dependency analysis was not possible. These warnings are suppressed by default and can be turned on by setting the flag, `SuppressTrimAnalysisWarnings`, to false. More information on the trim options available can be found at the [ILLinker page](https://github.com/mono/linker/blob/master/docs/illink-options.md).
```

> 

Reviewer 4: Fixes markdownlint error

```suggestion
```
